### ISSUE TITLE
[Question / friction] envoy unable to read /etc/envoy/envoy-config.json

### ISSUE THREAD

**Issue by cwbeitel** [kubeflow/kubeflow#2070]
Quick question about configuring envoy if anybody has a moment. Thought I followed instructions in https://www.kubeflow.org/docs/started/getting-started-gke/ pretty closely, here’s my exact deploy script https://gist.github.com/cwbeitel/dbee026c2e1f6852220b3580adaad6b7.

When I first tried to run the following:

```bash
${KUBEFLOW_SRC}/scripts/kfctl.sh init ${KFAPP} --platform gcp --project ${PROJECT}
cd ${KFAPP}
${KUBEFLOW_SRC}/scripts/kfctl.sh generate platform
${KUBEFLOW_SRC}/scripts/kfctl.sh apply platform
${KUBEFLOW_SRC}/scripts/kfctl.sh generate k8s
${KUBEFLOW_SRC}/scripts/kfctl.sh apply k8s
```

without a kubernetes cluster deployed I saw a timeout trying to reach the kube master (which I believe was left over from a previous deployment).

So I deployed a cluster with the following command

```bash

gcloud beta container clusters create kubeflow \
--disk-size=100 \
--disk-type=pd-ssd \
--enable-autoscaling --max-nodes=4 --min-nodes=2 \
--num-nodes=2 \
--region us-east1-d \
--scopes https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/devstorage.full_control \
--machine-type n1-highmem-8 \
--cluster-version 1.11.3-gke.18
```

Then after configuring RBAC I was able to run the kfctl.sh deployment nearly to completion except for the envoy pods being stuck on ContainerCreating, showing the following

```bash
Events:
  Type     Reason       Age                From                                              Message
  ----     ------       ----               ----                                              -------
  Normal   Scheduled    49s                default-scheduler                                 Successfully assigned kubeflow/envoy-69bf97959c-m2dpg to gke-kubeflow-default-pool-91b0292f-v84w
  Warning  FailedMount  15s (x7 over 47s)  kubelet, gke-kubeflow-default-pool-91b0292f-v84w  MountVolume.SetUp failed for volume "sa-key" : secrets "admin-gcp-sa" not found
```

To resolve this I added the section for creating service accounts manually to the deploy script provided at the start of this issue and the containers were able to be created but crash looped.

The following showed no problem (once a service account credential with the right permissions was used):

```bash

➜  ~ kubectl logs -n kubeflow envoy-69bf97959c-89wts iap   
Activated service account credentials for: [jlewi-kubeflow-rl@kubeflow-rl.iam.gserviceaccount.com]
[component_manager]
disable_update_check = true
[core]
account = jlewi-kubeflow-rl@kubeflow-rl.iam.gserviceaccount.com
disable_usage_reporting = true
project = kubeflow-rl
[metrics]
environment = github_docker_image

Your active configuration is: [default]
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
Waiting for backend id PROJECT=kubeflow-rl NAMESPACE=kubeflow SERVICE=envoy filter=name~k8s-be-32607-...
```

But the following still showed an error:

```
➜  ~ kubectl logs -n kubeflow envoy-69bf97959c-g59fn envoy
[2018-12-11 01:14:58.335][1][info][main] external/envoy/source/server/server.cc:184] initializing epoch 0 (hot restart version=9.200.16384.127.options=capacity=16384, num_slots=8209 hash=228984379728933363)
[2018-12-11 01:14:58.335][1][critical][main] external/envoy/source/server/server.cc:71] error initializing configuration '/etc/envoy/envoy-config.json': unable to read file: /etc/envoy/envoy-config.json
```

Ah it looks like @kunmingg observed this previously https://github.com/kubeflow/kubeflow/issues/1588. Perhaps this is the same issue that was fixed for other cases but not this current one. This is with v0.3.4.

**Comment by cwbeitel** [comment]
Upon doing `kubectl delete namespace kubeflow` and re-deploying a previous ksonnet app (v0.2.0) that has always worked without issue now jhub for this new deployment shows the same 401 unauthorized issue. I guess this implies the cluster has some faulty configuration outside of the kubeflow namespace?

**Comment by cwbeitel** [comment]
To I tore down and re-deployed the cluster and was observing the same issue and recreating the ksonnet app permitted v0.2.0 to deploy successfully (as previously). So perhaps the same issue could be resolved by re-trying the kfctl.sh command with a fresh ks app dir? And/or by finding a way around that first timeout trying to reach the kube api so kfctl can create the cluster itself (and perhaps in the process add some additional configuration)?

**Comment by cwbeitel** [comment]
I'm curious whether the issue is with [iap.libsonnet](https://github.com/kubeflow/kubeflow/blob/d758175b7c42587aa6efe65c39d4a9361bc1b9a9/kubeflow/gcp/iap.libsonnet#L181), should this be /var/envoy-config or /var/shared/envoy-config.json instead of /etc/envoy/envoy-config.json?

**Comment by jlewi** [comment]
@cwbeitel The GKE cluster is created when you do `kfctl apply platform`. So you should not be creating the cluster manually via `gcloud container clusters create`.

If you run the kfctl script and perform the platform and k8s steps it should create the GCP service accounts needed inside your cluster. If you don't observe that I would check the output of the script to see if any errors occurred.

The JupyterHub IAP unauthorized error happens because if you have IAP enabled port-forwarding to the JupyterHub pod won't work because your requests won't have a valid JWT in the header so the request will be rejected.

My suggestion would be to try doing a clean install with a new app name (to avoid conflicts with any current resources.

**Comment by cwbeitel** [comment]
Thanks for your response @jlewi!

I tried this again and was able to proceed through all of the documented [GKE deployment steps](https://www.kubeflow.org/docs/started/getting-started-gke/) (without manually deploying the cluster or creating service accounts) with the addition of fixes described [here](https://github.com/kubeflow/kubeflow/issues/1898) with the exception of a similar problem as described above with envoy. Logs from an envoy pod are [here](https://gist.github.com/cwbeitel/b7beb9d2cd534cfd7da856ade39bfa72).

I did observe an error in the process of running kfctl.sh:
```bash
ERROR: (gcloud.projects.set-iam-policy) User [myemailaddress@gmail.com] does not have permission to access project [kubeflow-rl:setIamPolicy] (or it may not exist): The caller does not have permission
INFO:root:Successfully patched IAM policy
+ gcloud --project=kubeflow-rl container clusters get-credentials --zone=us-east1-d clarify
Fetching cluster endpoint and auth data.
kubeconfig entry generated for clarify.
```
suggesting part of the problem is that the user running kfctl.sh does not have setIamPolicy permissions.

**Comment by cwbeitel** [comment]
This issue is resolved for me.

For future reference there were a few issues in addition to those covered in the troubleshooting guide and [this](https://github.com/kubeflow/kubeflow/issues/1898):

(1) As mentioned above, setIamPolicy permissions were necessary to properly configure service accounts, and
(2) the problem with envoy was coming from my project not having sufficient remaining quota for the number of backend services needed. Upon deleting one old backend service a portion of the envoy pods became available and upon deleting the cluster and several additional backend services and re-deploying all envoy services quickly became available.

To summarize it would be nice if deployment automation tools could alert users to issues such as these (quota or deployer role issues) to streamline problems solving.

/cc @jlewi I'll assume you want this closed and subsequent steps if any would be their own issues.
